subsys/net/lib/sockets/sockets_service.c

```diff
@@ -0,0 +1,245 @@
+#include <zephyr/posix/sys/eventfd.h>
+
+static int init_socket_service(void);
+static bool init_done;
+
+static K_MUTEX_DEFINE(lock);
+static K_CONDVAR_DEFINE(wait_start);
+
+static struct service {
+	struct zsock_pollfd *events;
+	int count;
+} ctx;
+
+void net_socket_service_foreach(net_socket_service_cb_t cb, void *user_data)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		cb(svc, user_data);
+	}
+}
+
+static void cleanup_svc_events(struct net_socket_service_desc *svc)
+{
+	for (int i = svc->idx; i < svc->idx + svc->pev_len; i++) {
+		for (int j = 0; j < svc->pev_len; j++) {
+			if (ctx.events[i].fd >= 0 &&
+			    ctx.events[i].fd == svc->pev[j].event.fd) {
+				ctx.events[i].fd = -1;
+				svc->pev[j].event.fd = -1;
+				svc->pev[j].event.events = 0;
+			}
+		}
+	}
+}
+
+int z_impl_net_socket_service_register(struct net_socket_service_desc *service,
+				       struct zsock_pollfd *fds, int len,
+				       void *user_data)
+{
+	int i, ret = -ENOENT;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	if (!init_done) {
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		if (svc == service) {
+			if (fds == NULL) {
+				cleanup_svc_events(svc);
+			} else {
+				if (len > svc->pev_len) {
+					NET_DBG("Too many file descriptors, max is %d for service %p",
+						svc->pev_len, svc);
+					ret = -ENOMEM;
+					break;
+				}
+
+				for (i = 0; i < len; i++) {
+					svc->pev[i].event = fds[i];
+					svc->pev[i].user_data = user_data;
+				}
+
+				for (i = svc->idx; i < svc->idx + svc->pev_len; i++) {
+					for (int j = 0; j < svc->pev_len; j++) {
+						ctx.events[i] = svc->pev[i].event;
+					}
+				}
+			}
+
+			/* Tell the thread to re-read the variables */
+			eventfd_write(ctx.events[0].fd, 1);
+			ret = 0;
+
+			break;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+static struct net_socket_service_desc *find_svc_and_event(
+	struct zsock_pollfd *pev,
+	struct net_socket_service_event **event)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		for (int i = 0; i < svc->pev_len; i++) {
+			if (svc->pev[i].event.fd == pev->fd) {
+				*event = &svc->pev[i];
+				return svc;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static int trigger_work(struct zsock_pollfd *pev)
+{
+	struct net_socket_service_event *event;
+	struct net_socket_service_desc *svc;
+	int ret;
+
+	svc = find_svc_and_event(pev, &event);
+	if (svc == NULL) {
+		return -ENOENT;
+	}
+
+	if (svc->work_q != NULL) {
+		ret = k_work_submit_to_queue(svc->work_q, &event->work);
+	} else {
+		ret = k_work_submit(&event->work);
+	}
+
+	return ret;
+}
+
+static void socket_service_thread(void)
+{
+	int ret, i, fd, count = 0;
+	eventfd_t value;
+
+	STRUCT_SECTION_COUNT(net_socket_service_desc, &ret);
+	if (ret == 0) {
+		NET_INFO("No socket services found, service disabled.");
+		goto fail;
+	}
+
+	/* Create contiguous k_poll_event array to enable socket polling */
```

Comment:
Is this comment in the right place? I don't see releavance to `k_poll_event` here.

Reply:
Yes, it was a leftover from previous version where I was planning to use `k_poll_event` directly, but that did not work as planned.